src/Chemistry.hpp

```diff
@@ -0,0 +1,141 @@
+#ifndef CHEMISTRY_HPP_ // NOLINT
+#define CHEMISTRY_HPP_
+//==============================================================================
+// TwoMomentRad - a radiation transport library for patch-based AMR codes
+// Copyright 2020 Benjamin Wibking.
+// Released under the MIT license. See LICENSE file included in the GitHub repo.
+//==============================================================================
+/// \file CloudyCooling.hpp
+/// \brief Defines methods for interpolating cooling rates from Cloudy tables.
+///
+
+#include <limits>
+
+#include "AMReX.H"
+#include "AMReX_BLassert.H"
+#include "AMReX_Extension.H"
+#include "AMReX_GpuQualifiers.H"
+
+#include "hydro_system.hpp"
+#include "radiation_system.hpp"
+
+#include "burn_type.H"
+#include "burner.H"
+#include "eos.H"
+#include "extern_parameters.H"
+
+namespace quokka::chemistry
+{
+template <typename problem_t> void computeChemistry(amrex::MultiFab &mf, const Real dt_in)
+{
+	BL_PROFILE("computeChemistry()")
+
+	const Real grav_constant = 6.674e-8;
+	const Real dt = dt_in;
+
+	const auto &ba = mf.boxArray();
+	const auto &dmap = mf.DistributionMap();
+	amrex::iMultiFab nsubstepsMF(ba, dmap, 1, 0);
+
+	for (amrex::MFIter iter(mf); iter.isValid(); ++iter) {
+		const amrex::Box &indexRange = iter.validbox();
+		auto const &state = mf.array(iter);
+		auto const &nsubsteps = nsubstepsMF.array(iter);
+
+		amrex::ParallelFor(indexRange, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+			const Real rho = state(i, j, k, HydroSystem<problem_t>::density_index);
+			const Real Eint = state(i, j, k, HydroSystem<problem_t>::internalEnergy_index);
+
+			Real chem[NumSpec] = {-1.0};
+			Real inmfracs[NumSpec] = {-1.0};
+
+			for (int nn = 0; nn < NumSpec; ++nn) {
+				chem[nn] = state(i, j, k, HydroSystem<problem_t>::scalar0_index + nn);
+			}
+
+			// do chemistry using microphysics
+			int nsteps = 1000;
+			// replace below with call to microphysics
+
+			burn_t chemstate;
+
+			for (int n = 0; n < nsteps; n++) {
+
+				for (int nn = 0; nn < NumSpec; ++nn) {
+					inmfracs[nn] = chem[nn] * rho / spmasses[nn];
+					chemstate.xn[nn] = inmfracs[nn];
+				}
+
+				// stop the test if dt is very small
+				if (dt < 10) {
+					break;
+				}
+
+				// stop the test if we have reached very high densities
+				if (rho > 3e-6) {
```

Comment:
This check does not make sense except within the Microphysics unit test.

Reply:
I think we should keep the density check. At densities > 3e-6, the chemistry is not going to be accurate. Even H2 cooling will not be accurate because H2 will be very optically thick and the approximation used by the current network will not give reasonable cooling rates.

Reply:
In that case, I think we should make it a user-supplied parameter, rather than a hard-coded value. We also need to have it abort the whole simulation, rather than just the integration for a single cell.

Reply:
The units of this value should also be documented.

Reply:
How do you abort the simulation? As in, what command do you use?

Reply:
Nvm, I am using `amrex::Abort`

Reply:
That should work.

Reply:
It gives this error:

```
/Users/psharda/Documents/quokka/src/Chemistry.hpp:65:18: warning: character constant too long for its type
/Users/psharda/Documents/quokka/src/Chemistry.hpp:65:5: error: no matching function for call to 'Abort'
                                amrex::Abort('Density exceeded 3e-6 g/cm^3!');
                                ^~~~~~~~~~~~
/Users/psharda/Documents/quokka/extern/amrex/Src/Base/AMReX.H:147:10: note: candidate function not viable: no known conversion from 'int' to 'const std::string' (aka 'const basic_string<char, char_traits<char>, allocator<char> >') for 1st argument
    void Abort (const std::string& msg);
         ^
/Users/psharda/Documents/quokka/extern/amrex/Src/Base/AMReX.H:152:10: note: candidate function not viable: no known conversion from 'int' to 'const char *' for 1st argument
    void Abort (const char * msg = nullptr) {
```

Is there an AMREX `.H` file is missing in `Chemistry.hpp`?

Reply:
Ah nvm, turns out I was using single quotes instead of double quotes for the abort message.